rke/schema_rke_cluster_services_kube_api.go

```diff
@@ -11,25 +11,111 @@ import (
 )
 
 const (
-	clusterServicesKubeAPIApiversionTag                      = "apiVersion"
-	clusterServicesKubeAPIKindTag                            = "kind"
-	clusterServicesKubeAPIAuditLogConfigPolicyAPIDefault     = "audit.k8s.io/v1"
-	clusterServicesKubeAPIEventRateLimitConfigAPIDefault     = "eventratelimit.admission.k8s.io/v1alpha1"
-	clusterServicesKubeAPISecretsEncryptionConfigAPIDefault  = "apiserver.config.k8s.io/v1"
-	clusterServicesKubeAPIAuditLogConfigPolicyKindDefault    = "Policy"
-	clusterServicesKubeAPIEventRateLimitConfigKindDefault    = "Configuration"
-	clusterServicesKubeAPISecretsEncryptionConfigKindDefault = "EncryptionConfiguration"
+	clusterServicesKubeAPIApiVersionTag                                = "apiVersion"
+	clusterServicesKubeAPIKindTag                                      = "kind"
+	clusterServicesKubeAPIAdmissionConfigurationAPIDefault             = "apiserver.config.k8s.io/v1"
+	clusterServicesKubeAPIAuditLogConfigPolicyAPIDefault               = "audit.k8s.io/v1"
+	clusterServicesKubeAPIEventRateLimitConfigAPIDefault               = "eventratelimit.admission.k8s.io/v1alpha1"
+	clusterServicesKubeAPISecretsEncryptionConfigAPIDefault            = "apiserver.config.k8s.io/v1"
+	clusterServicesKubeAPIAdmissionConfigurationPodSecurityAPIDefault  = "pod-security.admission.config.k8s.io/v1"
+	clusterServicesKubeAPIAdmissionConfigurationKindDefault            = "AdmissionConfiguration"
+	clusterServicesKubeAPIAuditLogConfigPolicyKindDefault              = "Policy"
+	clusterServicesKubeAPIEventRateLimitConfigKindDefault              = "Configuration"
+	clusterServicesKubeAPISecretsEncryptionConfigKindDefault           = "EncryptionConfiguration"
+	clusterServicesKubeAPIAdmissionConfigurationPodSecurityKindDefault = "PodSecurityConfiguration"
 )
 
 var (
 	clusterServicesKubeAPIRequired = []string{
-		clusterServicesKubeAPIApiversionTag,
+		clusterServicesKubeAPIApiVersionTag,
 		clusterServicesKubeAPIKindTag,
 	}
 )
 
 //Schemas
 
+func clusterRKEConfigServicesKubeAPIAdmissionConfigurationPluginsFields() map[string]*schema.Schema {
+	s := map[string]*schema.Schema{
+		"name": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Computed:    true,
+			Description: "Plugin name",
+		},
+		"path": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Description: "Plugin path",
+		},
+		"configuration": {
+			Type:     schema.TypeString,
+			Optional: true,
+			Computed: true,
+			ValidateFunc: func(val interface{}, key string) (warns []string, errs []error) {
+				v, ok := val.(string)
+				if !ok || len(v) == 0 {
+					return
+				}
```

Comment:
Should it return errors or fail if the `val` is not String?  
I am unsure if my question is valid because it relates to how the `ValidateFunc` works, which I may not be unaware of.

Reply:
Oh, this is old code from when I had designed it differently - sorry about that. `admission_configuration` is designed as a String now and there's the same pattern in the ValidateFunc for it. It should return an error if the length is zero yeah like admission config does here

```
if !ok || len(check) == 0 {
						errs = append(errs, fmt.Errorf("%s is required on yaml", k))
					}
```